{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updated property [core/project].\n",
      "Requirement already satisfied: pytz in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from -r ../../requirements.txt (line 1)) (2023.3.post1)\n",
      "Requirement already satisfied: google-cloud-pubsub in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from -r ../../requirements.txt (line 2)) (2.18.4)\n",
      "Requirement already satisfied: google-cloud-bigquery in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from -r ../../requirements.txt (line 3)) (3.13.0)\n",
      "Requirement already satisfied: apache-beam in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from -r ../../requirements.txt (line 4)) (2.51.0)\n",
      "Requirement already satisfied: anaconda in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from -r ../../requirements.txt (line 7)) (0.0.1.1)\n",
      "Requirement already satisfied: pandas in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from -r ../../requirements.txt (line 8)) (1.5.3)\n",
      "Requirement already satisfied: grpcio<2.0dev,>=1.51.3 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from google-cloud-pubsub->-r ../../requirements.txt (line 2)) (1.59.2)\n",
      "Requirement already satisfied: google-api-core!=2.0.*,!=2.1.*,!=2.10.*,!=2.2.*,!=2.3.*,!=2.4.*,!=2.5.*,!=2.6.*,!=2.7.*,!=2.8.*,!=2.9.*,<3.0.0dev,>=1.34.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from google-api-core[grpc]!=2.0.*,!=2.1.*,!=2.10.*,!=2.2.*,!=2.3.*,!=2.4.*,!=2.5.*,!=2.6.*,!=2.7.*,!=2.8.*,!=2.9.*,<3.0.0dev,>=1.34.0->google-cloud-pubsub->-r ../../requirements.txt (line 2)) (2.12.0)\n",
      "Requirement already satisfied: proto-plus<2.0.0dev,>=1.22.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from google-cloud-pubsub->-r ../../requirements.txt (line 2)) (1.22.3)\n",
      "Requirement already satisfied: protobuf!=3.20.0,!=3.20.1,!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<5.0.0dev,>=3.19.5 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from google-cloud-pubsub->-r ../../requirements.txt (line 2)) (4.24.4)\n",
      "Requirement already satisfied: grpc-google-iam-v1<1.0.0dev,>=0.12.4 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from google-cloud-pubsub->-r ../../requirements.txt (line 2)) (0.12.6)\n",
      "Requirement already satisfied: grpcio-status>=1.33.2 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from google-cloud-pubsub->-r ../../requirements.txt (line 2)) (1.59.2)\n",
      "Requirement already satisfied: google-cloud-core<3.0.0dev,>=1.6.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from google-cloud-bigquery->-r ../../requirements.txt (line 3)) (2.3.3)\n",
      "Requirement already satisfied: google-resumable-media<3.0dev,>=0.6.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from google-cloud-bigquery->-r ../../requirements.txt (line 3)) (2.6.0)\n",
      "Requirement already satisfied: packaging>=20.0.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from google-cloud-bigquery->-r ../../requirements.txt (line 3)) (23.2)\n",
      "Requirement already satisfied: python-dateutil<3.0dev,>=2.7.2 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from google-cloud-bigquery->-r ../../requirements.txt (line 3)) (2.8.2)\n",
      "Requirement already satisfied: requests<3.0.0dev,>=2.21.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from google-cloud-bigquery->-r ../../requirements.txt (line 3)) (2.31.0)\n",
      "Requirement already satisfied: crcmod<2.0,>=1.7 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam->-r ../../requirements.txt (line 4)) (1.7)\n",
      "Requirement already satisfied: orjson<4,>=3.9.7 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam->-r ../../requirements.txt (line 4)) (3.9.10)\n",
      "Requirement already satisfied: dill<0.3.2,>=0.3.1.1 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam->-r ../../requirements.txt (line 4)) (0.3.1.1)\n",
      "Requirement already satisfied: cloudpickle~=2.2.1 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam->-r ../../requirements.txt (line 4)) (2.2.1)\n",
      "Requirement already satisfied: fastavro<2,>=0.23.6 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam->-r ../../requirements.txt (line 4)) (1.9.0)\n",
      "Requirement already satisfied: fasteners<1.0,>=0.3 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam->-r ../../requirements.txt (line 4)) (0.19)\n",
      "Requirement already satisfied: hdfs<3.0.0,>=2.1.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam->-r ../../requirements.txt (line 4)) (2.7.3)\n",
      "Requirement already satisfied: httplib2<0.23.0,>=0.8 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam->-r ../../requirements.txt (line 4)) (0.22.0)\n",
      "Requirement already satisfied: js2py<1,>=0.74 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam->-r ../../requirements.txt (line 4)) (0.74)\n",
      "Requirement already satisfied: numpy<1.25.0,>=1.14.3 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam->-r ../../requirements.txt (line 4)) (1.24.4)\n",
      "Requirement already satisfied: objsize<0.7.0,>=0.6.1 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam->-r ../../requirements.txt (line 4)) (0.6.1)\n",
      "Requirement already satisfied: pymongo<5.0.0,>=3.8.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam->-r ../../requirements.txt (line 4)) (4.6.0)\n",
      "Requirement already satisfied: pydot<2,>=1.2.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam->-r ../../requirements.txt (line 4)) (1.4.2)\n",
      "Requirement already satisfied: regex>=2020.6.8 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam->-r ../../requirements.txt (line 4)) (2023.10.3)\n",
      "Requirement already satisfied: typing-extensions>=3.7.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam->-r ../../requirements.txt (line 4)) (4.8.0)\n",
      "Requirement already satisfied: zstandard<1,>=0.18.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam->-r ../../requirements.txt (line 4)) (0.22.0)\n",
      "Requirement already satisfied: pyarrow<12.0.0,>=3.0.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam->-r ../../requirements.txt (line 4)) (11.0.0)\n",
      "Requirement already satisfied: cachetools<6,>=3.1.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam[gcp]->-r ../../requirements.txt (line 5)) (5.3.2)\n",
      "Requirement already satisfied: google-apitools<0.5.32,>=0.5.31 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam[gcp]->-r ../../requirements.txt (line 5)) (0.5.31)\n",
      "Requirement already satisfied: google-auth<3,>=1.18.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam[gcp]->-r ../../requirements.txt (line 5)) (2.23.4)\n",
      "Requirement already satisfied: google-auth-httplib2<0.2.0,>=0.1.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam[gcp]->-r ../../requirements.txt (line 5)) (0.1.1)\n",
      "Requirement already satisfied: google-cloud-datastore<3,>=2.0.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam[gcp]->-r ../../requirements.txt (line 5)) (2.18.0)\n",
      "Requirement already satisfied: google-cloud-pubsublite<2,>=1.2.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam[gcp]->-r ../../requirements.txt (line 5)) (1.8.3)\n",
      "Requirement already satisfied: google-cloud-bigquery-storage<3,>=2.6.3 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam[gcp]->-r ../../requirements.txt (line 5)) (2.22.0)\n",
      "Requirement already satisfied: google-cloud-bigtable<3,>=2.19.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam[gcp]->-r ../../requirements.txt (line 5)) (2.21.0)\n",
      "Requirement already satisfied: google-cloud-spanner<4,>=3.0.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam[gcp]->-r ../../requirements.txt (line 5)) (3.40.1)\n",
      "Requirement already satisfied: google-cloud-dlp<4,>=3.0.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam[gcp]->-r ../../requirements.txt (line 5)) (3.13.0)\n",
      "Requirement already satisfied: google-cloud-language<3,>=2.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam[gcp]->-r ../../requirements.txt (line 5)) (2.11.1)\n",
      "Requirement already satisfied: google-cloud-videointelligence<3,>=2.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam[gcp]->-r ../../requirements.txt (line 5)) (2.11.4)\n",
      "Requirement already satisfied: google-cloud-vision<4,>=2 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam[gcp]->-r ../../requirements.txt (line 5)) (3.4.5)\n",
      "Requirement already satisfied: google-cloud-recommendations-ai<0.11.0,>=0.1.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam[gcp]->-r ../../requirements.txt (line 5)) (0.10.5)\n",
      "Requirement already satisfied: google-cloud-aiplatform<2.0,>=1.26.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam[gcp]->-r ../../requirements.txt (line 5)) (1.36.0)\n",
      "Requirement already satisfied: facets-overview<2,>=1.1.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam[interactive]->-r ../../requirements.txt (line 6)) (1.1.1)\n",
      "Requirement already satisfied: google-cloud-dataproc<6,>=5.0.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam[interactive]->-r ../../requirements.txt (line 6)) (5.7.0)\n",
      "Requirement already satisfied: ipython<9,>=8 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam[interactive]->-r ../../requirements.txt (line 6)) (8.17.2)\n",
      "Requirement already satisfied: ipykernel<7,>=6 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam[interactive]->-r ../../requirements.txt (line 6)) (6.26.0)\n",
      "Requirement already satisfied: ipywidgets<9,>=8 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam[interactive]->-r ../../requirements.txt (line 6)) (8.1.1)\n",
      "Requirement already satisfied: jupyter-client!=6.1.13,<8.2.1,>=6.1.11 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam[interactive]->-r ../../requirements.txt (line 6)) (8.2.0)\n",
      "Requirement already satisfied: timeloop<2,>=1.0.2 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam[interactive]->-r ../../requirements.txt (line 6)) (1.0.2)\n",
      "Requirement already satisfied: nbformat<6,>=5.0.5 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam[interactive]->-r ../../requirements.txt (line 6)) (5.9.2)\n",
      "Requirement already satisfied: nbconvert<8,>=6.2.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from apache-beam[interactive]->-r ../../requirements.txt (line 6)) (7.10.0)\n",
      "Requirement already satisfied: googleapis-common-protos<2.0.dev0,>=1.56.2 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from google-api-core!=2.0.*,!=2.1.*,!=2.10.*,!=2.2.*,!=2.3.*,!=2.4.*,!=2.5.*,!=2.6.*,!=2.7.*,!=2.8.*,!=2.9.*,<3.0.0dev,>=1.34.0->google-api-core[grpc]!=2.0.*,!=2.1.*,!=2.10.*,!=2.2.*,!=2.3.*,!=2.4.*,!=2.5.*,!=2.6.*,!=2.7.*,!=2.8.*,!=2.9.*,<3.0.0dev,>=1.34.0->google-cloud-pubsub->-r ../../requirements.txt (line 2)) (1.61.0)\n",
      "Requirement already satisfied: oauth2client>=1.4.12 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from google-apitools<0.5.32,>=0.5.31->apache-beam[gcp]->-r ../../requirements.txt (line 5)) (4.1.3)\n",
      "Requirement already satisfied: six>=1.12.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from google-apitools<0.5.32,>=0.5.31->apache-beam[gcp]->-r ../../requirements.txt (line 5)) (1.16.0)\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from google-auth<3,>=1.18.0->apache-beam[gcp]->-r ../../requirements.txt (line 5)) (0.3.0)\n",
      "Requirement already satisfied: rsa<5,>=3.1.4 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from google-auth<3,>=1.18.0->apache-beam[gcp]->-r ../../requirements.txt (line 5)) (4.9)\n",
      "Requirement already satisfied: google-cloud-storage<3.0.0dev,>=1.32.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from google-cloud-aiplatform<2.0,>=1.26.0->apache-beam[gcp]->-r ../../requirements.txt (line 5)) (2.13.0)\n",
      "Requirement already satisfied: google-cloud-resource-manager<3.0.0dev,>=1.3.3 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from google-cloud-aiplatform<2.0,>=1.26.0->apache-beam[gcp]->-r ../../requirements.txt (line 5)) (1.10.4)\n",
      "Requirement already satisfied: shapely<3.0.0dev in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from google-cloud-aiplatform<2.0,>=1.26.0->apache-beam[gcp]->-r ../../requirements.txt (line 5)) (2.0.2)\n",
      "Requirement already satisfied: overrides<7.0.0,>=6.0.1 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from google-cloud-pubsublite<2,>=1.2.0->apache-beam[gcp]->-r ../../requirements.txt (line 5)) (6.5.0)\n",
      "Requirement already satisfied: sqlparse>=0.4.4 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from google-cloud-spanner<4,>=3.0.0->apache-beam[gcp]->-r ../../requirements.txt (line 5)) (0.4.4)\n",
      "Requirement already satisfied: google-crc32c<2.0dev,>=1.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from google-resumable-media<3.0dev,>=0.6.0->google-cloud-bigquery->-r ../../requirements.txt (line 3)) (1.5.0)\n",
      "Requirement already satisfied: docopt in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from hdfs<3.0.0,>=2.1.0->apache-beam->-r ../../requirements.txt (line 4)) (0.6.2)\n",
      "Requirement already satisfied: pyparsing!=3.0.0,!=3.0.1,!=3.0.2,!=3.0.3,<4,>=2.4.2 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from httplib2<0.23.0,>=0.8->apache-beam->-r ../../requirements.txt (line 4)) (3.1.1)\n",
      "Requirement already satisfied: comm>=0.1.1 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from ipykernel<7,>=6->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (0.1.4)\n",
      "Requirement already satisfied: debugpy>=1.6.5 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from ipykernel<7,>=6->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (1.8.0)\n",
      "Requirement already satisfied: jupyter-core!=5.0.*,>=4.12 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from ipykernel<7,>=6->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (5.5.0)\n",
      "Requirement already satisfied: matplotlib-inline>=0.1 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from ipykernel<7,>=6->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (0.1.6)\n",
      "Requirement already satisfied: nest-asyncio in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from ipykernel<7,>=6->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (1.5.8)\n",
      "Requirement already satisfied: psutil in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from ipykernel<7,>=6->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (5.9.6)\n",
      "Requirement already satisfied: pyzmq>=20 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from ipykernel<7,>=6->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (25.1.1)\n",
      "Requirement already satisfied: tornado>=6.1 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from ipykernel<7,>=6->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (6.3.3)\n",
      "Requirement already satisfied: traitlets>=5.4.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from ipykernel<7,>=6->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (5.13.0)\n",
      "Requirement already satisfied: decorator in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from ipython<9,>=8->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (5.1.1)\n",
      "Requirement already satisfied: jedi>=0.16 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from ipython<9,>=8->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (0.19.1)\n",
      "Requirement already satisfied: prompt-toolkit!=3.0.37,<3.1.0,>=3.0.30 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from ipython<9,>=8->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (3.0.39)\n",
      "Requirement already satisfied: pygments>=2.4.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from ipython<9,>=8->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (2.16.1)\n",
      "Requirement already satisfied: stack-data in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from ipython<9,>=8->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (0.6.3)\n",
      "Requirement already satisfied: exceptiongroup in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from ipython<9,>=8->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (1.1.3)\n",
      "Requirement already satisfied: pexpect>4.3 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from ipython<9,>=8->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (4.8.0)\n",
      "Requirement already satisfied: widgetsnbextension~=4.0.9 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from ipywidgets<9,>=8->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (4.0.9)\n",
      "Requirement already satisfied: jupyterlab-widgets~=3.0.9 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from ipywidgets<9,>=8->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (3.0.9)\n",
      "Requirement already satisfied: tzlocal>=1.2 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from js2py<1,>=0.74->apache-beam->-r ../../requirements.txt (line 4)) (5.2)\n",
      "Requirement already satisfied: pyjsparser>=2.5.1 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from js2py<1,>=0.74->apache-beam->-r ../../requirements.txt (line 4)) (2.7.1)\n",
      "Requirement already satisfied: importlib-metadata>=4.8.3 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from jupyter-client!=6.1.13,<8.2.1,>=6.1.11->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (6.8.0)\n",
      "Requirement already satisfied: beautifulsoup4 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from nbconvert<8,>=6.2.0->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (4.12.2)\n",
      "Requirement already satisfied: bleach!=5.0.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from nbconvert<8,>=6.2.0->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (6.1.0)\n",
      "Requirement already satisfied: defusedxml in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from nbconvert<8,>=6.2.0->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (0.7.1)\n",
      "Requirement already satisfied: jinja2>=3.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from nbconvert<8,>=6.2.0->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (3.1.2)\n",
      "Requirement already satisfied: jupyterlab-pygments in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from nbconvert<8,>=6.2.0->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (0.2.2)\n",
      "Requirement already satisfied: markupsafe>=2.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from nbconvert<8,>=6.2.0->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (2.1.3)\n",
      "Requirement already satisfied: mistune<4,>=2.0.3 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from nbconvert<8,>=6.2.0->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (3.0.2)\n",
      "Requirement already satisfied: nbclient>=0.5.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from nbconvert<8,>=6.2.0->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (0.8.0)\n",
      "Requirement already satisfied: pandocfilters>=1.4.1 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from nbconvert<8,>=6.2.0->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (1.5.0)\n",
      "Requirement already satisfied: tinycss2 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from nbconvert<8,>=6.2.0->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (1.2.1)\n",
      "Requirement already satisfied: fastjsonschema in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from nbformat<6,>=5.0.5->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (2.18.1)\n",
      "Requirement already satisfied: jsonschema>=2.6 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from nbformat<6,>=5.0.5->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (4.19.2)\n",
      "Requirement already satisfied: dnspython<3.0.0,>=1.16.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from pymongo<5.0.0,>=3.8.0->apache-beam->-r ../../requirements.txt (line 4)) (2.4.2)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from requests<3.0.0dev,>=2.21.0->google-cloud-bigquery->-r ../../requirements.txt (line 3)) (3.3.2)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from requests<3.0.0dev,>=2.21.0->google-cloud-bigquery->-r ../../requirements.txt (line 3)) (3.4)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from requests<3.0.0dev,>=2.21.0->google-cloud-bigquery->-r ../../requirements.txt (line 3)) (2.0.7)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from requests<3.0.0dev,>=2.21.0->google-cloud-bigquery->-r ../../requirements.txt (line 3)) (2023.7.22)\n",
      "Requirement already satisfied: webencodings in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from bleach!=5.0.0->nbconvert<8,>=6.2.0->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (0.5.1)\n",
      "Requirement already satisfied: zipp>=0.5 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from importlib-metadata>=4.8.3->jupyter-client!=6.1.13,<8.2.1,>=6.1.11->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (3.17.0)\n",
      "Requirement already satisfied: parso<0.9.0,>=0.8.3 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from jedi>=0.16->ipython<9,>=8->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (0.8.3)\n",
      "Requirement already satisfied: attrs>=22.2.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from jsonschema>=2.6->nbformat<6,>=5.0.5->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (23.1.0)\n",
      "Requirement already satisfied: jsonschema-specifications>=2023.03.6 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from jsonschema>=2.6->nbformat<6,>=5.0.5->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (2023.7.1)\n",
      "Requirement already satisfied: referencing>=0.28.4 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from jsonschema>=2.6->nbformat<6,>=5.0.5->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (0.30.2)\n",
      "Requirement already satisfied: rpds-py>=0.7.1 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from jsonschema>=2.6->nbformat<6,>=5.0.5->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (0.12.0)\n",
      "Requirement already satisfied: platformdirs>=2.5 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from jupyter-core!=5.0.*,>=4.12->ipykernel<7,>=6->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (3.11.0)\n",
      "Requirement already satisfied: pyasn1>=0.1.7 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from oauth2client>=1.4.12->google-apitools<0.5.32,>=0.5.31->apache-beam[gcp]->-r ../../requirements.txt (line 5)) (0.5.0)\n",
      "Requirement already satisfied: ptyprocess>=0.5 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from pexpect>4.3->ipython<9,>=8->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (0.7.0)\n",
      "Requirement already satisfied: wcwidth in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from prompt-toolkit!=3.0.37,<3.1.0,>=3.0.30->ipython<9,>=8->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (0.2.9)\n",
      "Requirement already satisfied: soupsieve>1.2 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from beautifulsoup4->nbconvert<8,>=6.2.0->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (2.5)\n",
      "Requirement already satisfied: executing>=1.2.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from stack-data->ipython<9,>=8->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (2.0.1)\n",
      "Requirement already satisfied: asttokens>=2.1.0 in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from stack-data->ipython<9,>=8->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (2.4.1)\n",
      "Requirement already satisfied: pure-eval in /home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages (from stack-data->ipython<9,>=8->apache-beam[interactive]->-r ../../requirements.txt (line 6)) (0.2.2)\n"
     ]
    }
   ],
   "source": [
    "!{'gcloud config set project poc-dbs'}\n",
    "!{'pip install -r ../../requirements.txt'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import apache_beam as beam\n",
    "import logging\n",
    "import json\n",
    "import numpy as np\n",
    "\n",
    "DATETIME_FORMAT = '%Y-%m-%dT%H:%M:%S'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_stats(airport, events):\n",
    "    arrived = [event['ARR_DELAY'] for event in events if event['EVENT_TYPE'] == 'arrived']\n",
    "    avg_arr_delay = float(np.mean(arrived)) if len(arrived) > 0 else None\n",
    "\n",
    "    departed = [event['DEP_DELAY'] for event in events if event['EVENT_TYPE'] == 'departed']\n",
    "    avg_dep_delay = float(np.mean(departed)) if len(departed) > 0 else None\n",
    "\n",
    "    num_flights = len(events)\n",
    "    start_time = min([event['EVENT_TIME'] for event in events])\n",
    "    latest_time = max([event['EVENT_TIME'] for event in events])\n",
    "\n",
    "    return {\n",
    "        'AIRPORT': airport,\n",
    "        'AVG_ARR_DELAY': avg_arr_delay,\n",
    "        'AVG_DEP_DELAY': avg_dep_delay,\n",
    "        'NUM_FLIGHTS': num_flights,\n",
    "        'START_TIME': start_time,\n",
    "        'END_TIME': latest_time\n",
    "    }\n",
    "\n",
    "\n",
    "def by_airport(event):\n",
    "    if event['EVENT_TYPE'] == 'departed':\n",
    "        return event['ORIGIN'], event\n",
    "    else:\n",
    "        return event['DEST'], event\n",
    "\n",
    "project='poc-dbs'\n",
    "bucket='poc-dbs-dsongcp'\n",
    "region='europe-west4'\n",
    "\n",
    "argv = [\n",
    "    '--project={0}'.format(project),\n",
    "    '--job_name=ch04avgdelay',\n",
    "    '--streaming',\n",
    "    '--save_main_session',\n",
    "    '--staging_location=gs://{0}/flights/staging/'.format(bucket),\n",
    "    '--temp_location=gs://{0}/flights/temp/'.format(bucket),\n",
    "    '--autoscaling_algorithm=THROUGHPUT_BASED',\n",
    "    '--max_num_workers=8',\n",
    "    '--region={}'.format(region),\n",
    "    '--runner=DirectRunner'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updated property [core/project].\n"
     ]
    },
    {
     "data": {
      "application/javascript": "\n        if (typeof window.interactive_beam_jquery == 'undefined') {\n          var jqueryScript = document.createElement('script');\n          jqueryScript.src = 'https://code.jquery.com/jquery-3.4.1.slim.min.js';\n          jqueryScript.type = 'text/javascript';\n          jqueryScript.onload = function() {\n            var datatableScript = document.createElement('script');\n            datatableScript.src = 'https://cdn.datatables.net/1.10.20/js/jquery.dataTables.min.js';\n            datatableScript.type = 'text/javascript';\n            datatableScript.onload = function() {\n              window.interactive_beam_jquery = jQuery.noConflict(true);\n              window.interactive_beam_jquery(document).ready(function($){\n                \n              });\n            }\n            document.head.appendChild(datatableScript);\n          };\n          document.head.appendChild(jqueryScript);\n        } else {\n          window.interactive_beam_jquery(document).ready(function($){\n            \n          });\n        }"
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m/home/antoine_baret/data-science-on-gcp/04_streaming/realtime/notebook.ipynb Cell 4\u001b[0m line \u001b[0;36m2\n\u001b[1;32m     <a href='vscode-notebook-cell://970-cs-1053622369794-default.cs-europe-west1-iuzs.cloudshell.dev/home/antoine_baret/data-science-on-gcp/04_streaming/realtime/notebook.ipynb#W3sdnNjb2RlLXJlbW90ZQ%3D%3D?line=14'>15</a>\u001b[0m all_events \u001b[39m=\u001b[39m (events[\u001b[39m'\u001b[39m\u001b[39marrived\u001b[39m\u001b[39m'\u001b[39m], events[\u001b[39m'\u001b[39m\u001b[39mdeparted\u001b[39m\u001b[39m'\u001b[39m]) \u001b[39m|\u001b[39m beam\u001b[39m.\u001b[39mFlatten()\n\u001b[1;32m     <a href='vscode-notebook-cell://970-cs-1053622369794-default.cs-europe-west1-iuzs.cloudshell.dev/home/antoine_baret/data-science-on-gcp/04_streaming/realtime/notebook.ipynb#W3sdnNjb2RlLXJlbW90ZQ%3D%3D?line=16'>17</a>\u001b[0m stats \u001b[39m=\u001b[39m (all_events\n\u001b[1;32m     <a href='vscode-notebook-cell://970-cs-1053622369794-default.cs-europe-west1-iuzs.cloudshell.dev/home/antoine_baret/data-science-on-gcp/04_streaming/realtime/notebook.ipynb#W3sdnNjb2RlLXJlbW90ZQ%3D%3D?line=17'>18</a>\u001b[0m          \u001b[39m|\u001b[39m \u001b[39m'\u001b[39m\u001b[39mbyairport\u001b[39m\u001b[39m'\u001b[39m \u001b[39m>>\u001b[39m beam\u001b[39m.\u001b[39mMap(by_airport)\n\u001b[1;32m     <a href='vscode-notebook-cell://970-cs-1053622369794-default.cs-europe-west1-iuzs.cloudshell.dev/home/antoine_baret/data-science-on-gcp/04_streaming/realtime/notebook.ipynb#W3sdnNjb2RlLXJlbW90ZQ%3D%3D?line=18'>19</a>\u001b[0m          \u001b[39m|\u001b[39m \u001b[39m'\u001b[39m\u001b[39mwindow\u001b[39m\u001b[39m'\u001b[39m \u001b[39m>>\u001b[39m beam\u001b[39m.\u001b[39mWindowInto(beam\u001b[39m.\u001b[39mwindow\u001b[39m.\u001b[39mSlidingWindows(\u001b[39m60\u001b[39m \u001b[39m*\u001b[39m \u001b[39m60\u001b[39m, \u001b[39m5\u001b[39m \u001b[39m*\u001b[39m \u001b[39m60\u001b[39m))\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     <a href='vscode-notebook-cell://970-cs-1053622369794-default.cs-europe-west1-iuzs.cloudshell.dev/home/antoine_baret/data-science-on-gcp/04_streaming/realtime/notebook.ipynb#W3sdnNjb2RlLXJlbW90ZQ%3D%3D?line=21'>22</a>\u001b[0m          \u001b[39m|\u001b[39m beam\u001b[39m.\u001b[39mMap(\u001b[39mlambda\u001b[39;00m x: \u001b[39mprint\u001b[39m(x))\n\u001b[1;32m     <a href='vscode-notebook-cell://970-cs-1053622369794-default.cs-europe-west1-iuzs.cloudshell.dev/home/antoine_baret/data-science-on-gcp/04_streaming/realtime/notebook.ipynb#W3sdnNjb2RlLXJlbW90ZQ%3D%3D?line=22'>23</a>\u001b[0m )\n\u001b[0;32m---> <a href='vscode-notebook-cell://970-cs-1053622369794-default.cs-europe-west1-iuzs.cloudshell.dev/home/antoine_baret/data-science-on-gcp/04_streaming/realtime/notebook.ipynb#W3sdnNjb2RlLXJlbW90ZQ%3D%3D?line=25'>26</a>\u001b[0m \u001b[39m\u001b[39m\u001b[39m\"\"\"\u001b[39;00m\n\u001b[1;32m     <a href='vscode-notebook-cell://970-cs-1053622369794-default.cs-europe-west1-iuzs.cloudshell.dev/home/antoine_baret/data-science-on-gcp/04_streaming/realtime/notebook.ipynb#W3sdnNjb2RlLXJlbW90ZQ%3D%3D?line=26'>27</a>\u001b[0m \u001b[39mstats_schema = ','.join(['AIRPORT:string,AVG_ARR_DELAY:float,AVG_DEP_DELAY:float',\u001b[39;00m\n\u001b[1;32m     <a href='vscode-notebook-cell://970-cs-1053622369794-default.cs-europe-west1-iuzs.cloudshell.dev/home/antoine_baret/data-science-on-gcp/04_streaming/realtime/notebook.ipynb#W3sdnNjb2RlLXJlbW90ZQ%3D%3D?line=27'>28</a>\u001b[0m \u001b[39m                         'NUM_FLIGHTS:int64,START_TIME:timestamp,END_TIME:timestamp']\u001b[39;00m\n\u001b[1;32m     <a href='vscode-notebook-cell://970-cs-1053622369794-default.cs-europe-west1-iuzs.cloudshell.dev/home/antoine_baret/data-science-on-gcp/04_streaming/realtime/notebook.ipynb#W3sdnNjb2RlLXJlbW90ZQ%3D%3D?line=28'>29</a>\u001b[0m \u001b[39m                         )\u001b[39;00m\n\u001b[1;32m     <a href='vscode-notebook-cell://970-cs-1053622369794-default.cs-europe-west1-iuzs.cloudshell.dev/home/antoine_baret/data-science-on-gcp/04_streaming/realtime/notebook.ipynb#W3sdnNjb2RlLXJlbW90ZQ%3D%3D?line=29'>30</a>\u001b[0m \u001b[39m(stats\u001b[39;00m\n\u001b[1;32m     <a href='vscode-notebook-cell://970-cs-1053622369794-default.cs-europe-west1-iuzs.cloudshell.dev/home/antoine_baret/data-science-on-gcp/04_streaming/realtime/notebook.ipynb#W3sdnNjb2RlLXJlbW90ZQ%3D%3D?line=30'>31</a>\u001b[0m \u001b[39m | 'bqout' >> beam.io.gcp.bigquery.WriteToBigQuery(\u001b[39;00m\n\u001b[1;32m     <a href='vscode-notebook-cell://970-cs-1053622369794-default.cs-europe-west1-iuzs.cloudshell.dev/home/antoine_baret/data-science-on-gcp/04_streaming/realtime/notebook.ipynb#W3sdnNjb2RlLXJlbW90ZQ%3D%3D?line=31'>32</a>\u001b[0m \u001b[39m            table='dsongcp.streaming_delays', \u001b[39;00m\n\u001b[1;32m     <a href='vscode-notebook-cell://970-cs-1053622369794-default.cs-europe-west1-iuzs.cloudshell.dev/home/antoine_baret/data-science-on-gcp/04_streaming/realtime/notebook.ipynb#W3sdnNjb2RlLXJlbW90ZQ%3D%3D?line=32'>33</a>\u001b[0m \u001b[39m            schema=stats_schema\u001b[39;00m\n\u001b[1;32m     <a href='vscode-notebook-cell://970-cs-1053622369794-default.cs-europe-west1-iuzs.cloudshell.dev/home/antoine_baret/data-science-on-gcp/04_streaming/realtime/notebook.ipynb#W3sdnNjb2RlLXJlbW90ZQ%3D%3D?line=33'>34</a>\u001b[0m \u001b[39m        )\u001b[39;00m\n\u001b[1;32m     <a href='vscode-notebook-cell://970-cs-1053622369794-default.cs-europe-west1-iuzs.cloudshell.dev/home/antoine_baret/data-science-on-gcp/04_streaming/realtime/notebook.ipynb#W3sdnNjb2RlLXJlbW90ZQ%3D%3D?line=34'>35</a>\u001b[0m \u001b[39m )\"\"\"\u001b[39;00m\n",
      "File \u001b[0;32m~/data-science-on-gcp/.venv/lib/python3.9/site-packages/apache_beam/pipeline.py:601\u001b[0m, in \u001b[0;36mPipeline.__exit__\u001b[0;34m(self, exc_type, exc_val, exc_tb)\u001b[0m\n\u001b[1;32m    599\u001b[0m   \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m exc_type:\n\u001b[1;32m    600\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mresult \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mrun()\n\u001b[0;32m--> 601\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mresult\u001b[39m.\u001b[39;49mwait_until_finish()\n\u001b[1;32m    602\u001b[0m \u001b[39mfinally\u001b[39;00m:\n\u001b[1;32m    603\u001b[0m   \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_extra_context\u001b[39m.\u001b[39m\u001b[39m__exit__\u001b[39m(exc_type, exc_val, exc_tb)\n",
      "File \u001b[0;32m~/data-science-on-gcp/.venv/lib/python3.9/site-packages/apache_beam/runners/direct/direct_runner.py:585\u001b[0m, in \u001b[0;36mDirectPipelineResult.wait_until_finish\u001b[0;34m(self, duration)\u001b[0m\n\u001b[1;32m    582\u001b[0m   \u001b[39mraise\u001b[39;00m \u001b[39mNotImplementedError\u001b[39;00m(\n\u001b[1;32m    583\u001b[0m       \u001b[39m'\u001b[39m\u001b[39mDirectRunner does not support duration argument.\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[1;32m    584\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 585\u001b[0m   \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_executor\u001b[39m.\u001b[39;49mawait_completion()\n\u001b[1;32m    586\u001b[0m   \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_state \u001b[39m=\u001b[39m PipelineState\u001b[39m.\u001b[39mDONE\n\u001b[1;32m    587\u001b[0m \u001b[39mexcept\u001b[39;00m:  \u001b[39m# pylint: disable=broad-except\u001b[39;00m\n",
      "File \u001b[0;32m~/data-science-on-gcp/.venv/lib/python3.9/site-packages/apache_beam/runners/direct/executor.py:432\u001b[0m, in \u001b[0;36mExecutor.await_completion\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    431\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mawait_completion\u001b[39m(\u001b[39mself\u001b[39m):\n\u001b[0;32m--> 432\u001b[0m   \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_executor\u001b[39m.\u001b[39;49mawait_completion()\n",
      "File \u001b[0;32m~/data-science-on-gcp/.venv/lib/python3.9/site-packages/apache_beam/runners/direct/executor.py:477\u001b[0m, in \u001b[0;36m_ExecutorServiceParallelExecutor.await_completion\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    476\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mawait_completion\u001b[39m(\u001b[39mself\u001b[39m):\n\u001b[0;32m--> 477\u001b[0m   update \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mvisible_updates\u001b[39m.\u001b[39;49mtake()\n\u001b[1;32m    478\u001b[0m   \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m    479\u001b[0m     \u001b[39mif\u001b[39;00m update\u001b[39m.\u001b[39mexception:\n",
      "File \u001b[0;32m~/data-science-on-gcp/.venv/lib/python3.9/site-packages/apache_beam/runners/direct/executor.py:551\u001b[0m, in \u001b[0;36m_ExecutorServiceParallelExecutor._TypedUpdateQueue.take\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    549\u001b[0m \u001b[39mwhile\u001b[39;00m \u001b[39mTrue\u001b[39;00m:\n\u001b[1;32m    550\u001b[0m   \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 551\u001b[0m     item \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_queue\u001b[39m.\u001b[39;49mget(timeout\u001b[39m=\u001b[39;49m\u001b[39m1\u001b[39;49m)\n\u001b[1;32m    552\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_queue\u001b[39m.\u001b[39mtask_done()\n\u001b[1;32m    553\u001b[0m     \u001b[39mreturn\u001b[39;00m item\n",
      "File \u001b[0;32m/usr/lib/python3.9/queue.py:180\u001b[0m, in \u001b[0;36mQueue.get\u001b[0;34m(self, block, timeout)\u001b[0m\n\u001b[1;32m    178\u001b[0m         \u001b[39mif\u001b[39;00m remaining \u001b[39m<\u001b[39m\u001b[39m=\u001b[39m \u001b[39m0.0\u001b[39m:\n\u001b[1;32m    179\u001b[0m             \u001b[39mraise\u001b[39;00m Empty\n\u001b[0;32m--> 180\u001b[0m         \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mnot_empty\u001b[39m.\u001b[39;49mwait(remaining)\n\u001b[1;32m    181\u001b[0m item \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_get()\n\u001b[1;32m    182\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mnot_full\u001b[39m.\u001b[39mnotify()\n",
      "File \u001b[0;32m/usr/lib/python3.9/threading.py:316\u001b[0m, in \u001b[0;36mCondition.wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    314\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    315\u001b[0m     \u001b[39mif\u001b[39;00m timeout \u001b[39m>\u001b[39m \u001b[39m0\u001b[39m:\n\u001b[0;32m--> 316\u001b[0m         gotit \u001b[39m=\u001b[39m waiter\u001b[39m.\u001b[39;49macquire(\u001b[39mTrue\u001b[39;49;00m, timeout)\n\u001b[1;32m    317\u001b[0m     \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    318\u001b[0m         gotit \u001b[39m=\u001b[39m waiter\u001b[39m.\u001b[39macquire(\u001b[39mFalse\u001b[39;00m)\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "ERROR:apache_beam.runners.common:unsupported operand type(s) for /: 'NoneType' and 'int' [while running '[4]: stats']\n",
      "Traceback (most recent call last):\n",
      "  File \"apache_beam/runners/common.py\", line 1434, in apache_beam.runners.common.DoFnRunner.process\n",
      "  File \"apache_beam/runners/common.py\", line 637, in apache_beam.runners.common.SimpleInvoker.invoke_process\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/apache_beam/transforms/core.py\", line 1963, in <lambda>\n",
      "    wrapper = lambda x: [fn(x)]\n",
      "  File \"/tmp/ipykernel_3858/2327646342.py\", line 21, in <lambda>\n",
      "    | 'stats' >> beam.Map(lambda x: compute_stats(x[0], x[1]))\n",
      "  File \"/tmp/ipykernel_3858/3647296015.py\", line 3, in compute_stats\n",
      "    avg_arr_delay = float(np.mean(arrived)) if len(arrived) > 0 else None\n",
      "  File \"<__array_function__ internals>\", line 200, in mean\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/numpy/core/fromnumeric.py\", line 3464, in mean\n",
      "    return _methods._mean(a, axis=axis, dtype=dtype,\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/numpy/core/_methods.py\", line 194, in _mean\n",
      "    ret = ret / rcount\n",
      "TypeError: unsupported operand type(s) for /: 'NoneType' and 'int'\n",
      "ERROR:apache_beam.runners.direct.executor:Exception at bundle <apache_beam.runners.direct.bundle_factory._Bundle object at 0x78d32553f180>, due to an exception.\n",
      " Traceback (most recent call last):\n",
      "  File \"apache_beam/runners/common.py\", line 1434, in apache_beam.runners.common.DoFnRunner.process\n",
      "  File \"apache_beam/runners/common.py\", line 637, in apache_beam.runners.common.SimpleInvoker.invoke_process\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/apache_beam/transforms/core.py\", line 1963, in <lambda>\n",
      "    wrapper = lambda x: [fn(x)]\n",
      "  File \"/tmp/ipykernel_3858/2327646342.py\", line 21, in <lambda>\n",
      "    | 'stats' >> beam.Map(lambda x: compute_stats(x[0], x[1]))\n",
      "  File \"/tmp/ipykernel_3858/3647296015.py\", line 3, in compute_stats\n",
      "    avg_arr_delay = float(np.mean(arrived)) if len(arrived) > 0 else None\n",
      "  File \"<__array_function__ internals>\", line 200, in mean\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/numpy/core/fromnumeric.py\", line 3464, in mean\n",
      "    return _methods._mean(a, axis=axis, dtype=dtype,\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/numpy/core/_methods.py\", line 194, in _mean\n",
      "    ret = ret / rcount\n",
      "TypeError: unsupported operand type(s) for /: 'NoneType' and 'int'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/apache_beam/runners/direct/executor.py\", line 370, in call\n",
      "    self.attempt_call(\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/apache_beam/runners/direct/executor.py\", line 413, in attempt_call\n",
      "    evaluator.process_element(value)\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 883, in process_element\n",
      "    self.runner.process(element)\n",
      "  File \"apache_beam/runners/common.py\", line 1431, in apache_beam.runners.common.DoFnRunner.process\n",
      "  File \"apache_beam/runners/common.py\", line 1436, in apache_beam.runners.common.DoFnRunner.process\n",
      "  File \"apache_beam/runners/common.py\", line 1545, in apache_beam.runners.common.DoFnRunner._reraise_augmented\n",
      "  File \"apache_beam/runners/common.py\", line 1434, in apache_beam.runners.common.DoFnRunner.process\n",
      "  File \"apache_beam/runners/common.py\", line 637, in apache_beam.runners.common.SimpleInvoker.invoke_process\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/apache_beam/transforms/core.py\", line 1963, in <lambda>\n",
      "    wrapper = lambda x: [fn(x)]\n",
      "  File \"/tmp/ipykernel_3858/2327646342.py\", line 21, in <lambda>\n",
      "    | 'stats' >> beam.Map(lambda x: compute_stats(x[0], x[1]))\n",
      "  File \"/tmp/ipykernel_3858/3647296015.py\", line 3, in compute_stats\n",
      "    avg_arr_delay = float(np.mean(arrived)) if len(arrived) > 0 else None\n",
      "  File \"<__array_function__ internals>\", line 200, in mean\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/numpy/core/fromnumeric.py\", line 3464, in mean\n",
      "    return _methods._mean(a, axis=axis, dtype=dtype,\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/numpy/core/_methods.py\", line 194, in _mean\n",
      "    ret = ret / rcount\n",
      "TypeError: unsupported operand type(s) for /: 'NoneType' and 'int' [while running '[4]: stats']\n",
      "\n",
      "ERROR:apache_beam.runners.common:unsupported operand type(s) for /: 'NoneType' and 'int' [while running '[4]: stats']\n",
      "Traceback (most recent call last):\n",
      "  File \"apache_beam/runners/common.py\", line 1434, in apache_beam.runners.common.DoFnRunner.process\n",
      "  File \"apache_beam/runners/common.py\", line 637, in apache_beam.runners.common.SimpleInvoker.invoke_process\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/apache_beam/transforms/core.py\", line 1963, in <lambda>\n",
      "    wrapper = lambda x: [fn(x)]\n",
      "  File \"/tmp/ipykernel_3858/2327646342.py\", line 21, in <lambda>\n",
      "    | 'stats' >> beam.Map(lambda x: compute_stats(x[0], x[1]))\n",
      "  File \"/tmp/ipykernel_3858/3647296015.py\", line 3, in compute_stats\n",
      "    avg_arr_delay = float(np.mean(arrived)) if len(arrived) > 0 else None\n",
      "  File \"<__array_function__ internals>\", line 200, in mean\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/numpy/core/fromnumeric.py\", line 3464, in mean\n",
      "    return _methods._mean(a, axis=axis, dtype=dtype,\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/numpy/core/_methods.py\", line 194, in _mean\n",
      "    ret = ret / rcount\n",
      "TypeError: unsupported operand type(s) for /: 'NoneType' and 'int'\n",
      "ERROR:apache_beam.runners.direct.executor:Exception at bundle <apache_beam.runners.direct.bundle_factory._Bundle object at 0x78d32553f180>, due to an exception.\n",
      " Traceback (most recent call last):\n",
      "  File \"apache_beam/runners/common.py\", line 1434, in apache_beam.runners.common.DoFnRunner.process\n",
      "  File \"apache_beam/runners/common.py\", line 637, in apache_beam.runners.common.SimpleInvoker.invoke_process\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/apache_beam/transforms/core.py\", line 1963, in <lambda>\n",
      "    wrapper = lambda x: [fn(x)]\n",
      "  File \"/tmp/ipykernel_3858/2327646342.py\", line 21, in <lambda>\n",
      "    | 'stats' >> beam.Map(lambda x: compute_stats(x[0], x[1]))\n",
      "  File \"/tmp/ipykernel_3858/3647296015.py\", line 3, in compute_stats\n",
      "    avg_arr_delay = float(np.mean(arrived)) if len(arrived) > 0 else None\n",
      "  File \"<__array_function__ internals>\", line 200, in mean\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/numpy/core/fromnumeric.py\", line 3464, in mean\n",
      "    return _methods._mean(a, axis=axis, dtype=dtype,\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/numpy/core/_methods.py\", line 194, in _mean\n",
      "    ret = ret / rcount\n",
      "TypeError: unsupported operand type(s) for /: 'NoneType' and 'int'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/apache_beam/runners/direct/executor.py\", line 370, in call\n",
      "    self.attempt_call(\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/apache_beam/runners/direct/executor.py\", line 413, in attempt_call\n",
      "    evaluator.process_element(value)\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 883, in process_element\n",
      "    self.runner.process(element)\n",
      "  File \"apache_beam/runners/common.py\", line 1431, in apache_beam.runners.common.DoFnRunner.process\n",
      "  File \"apache_beam/runners/common.py\", line 1436, in apache_beam.runners.common.DoFnRunner.process\n",
      "  File \"apache_beam/runners/common.py\", line 1545, in apache_beam.runners.common.DoFnRunner._reraise_augmented\n",
      "  File \"apache_beam/runners/common.py\", line 1434, in apache_beam.runners.common.DoFnRunner.process\n",
      "  File \"apache_beam/runners/common.py\", line 637, in apache_beam.runners.common.SimpleInvoker.invoke_process\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/apache_beam/transforms/core.py\", line 1963, in <lambda>\n",
      "    wrapper = lambda x: [fn(x)]\n",
      "  File \"/tmp/ipykernel_3858/2327646342.py\", line 21, in <lambda>\n",
      "    | 'stats' >> beam.Map(lambda x: compute_stats(x[0], x[1]))\n",
      "  File \"/tmp/ipykernel_3858/3647296015.py\", line 3, in compute_stats\n",
      "    avg_arr_delay = float(np.mean(arrived)) if len(arrived) > 0 else None\n",
      "  File \"<__array_function__ internals>\", line 200, in mean\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/numpy/core/fromnumeric.py\", line 3464, in mean\n",
      "    return _methods._mean(a, axis=axis, dtype=dtype,\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/numpy/core/_methods.py\", line 194, in _mean\n",
      "    ret = ret / rcount\n",
      "TypeError: unsupported operand type(s) for /: 'NoneType' and 'int' [while running '[4]: stats']\n",
      "\n",
      "ERROR:apache_beam.runners.common:unsupported operand type(s) for /: 'NoneType' and 'int' [while running '[4]: stats']\n",
      "Traceback (most recent call last):\n",
      "  File \"apache_beam/runners/common.py\", line 1434, in apache_beam.runners.common.DoFnRunner.process\n",
      "  File \"apache_beam/runners/common.py\", line 637, in apache_beam.runners.common.SimpleInvoker.invoke_process\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/apache_beam/transforms/core.py\", line 1963, in <lambda>\n",
      "    wrapper = lambda x: [fn(x)]\n",
      "  File \"/tmp/ipykernel_3858/2327646342.py\", line 21, in <lambda>\n",
      "    | 'stats' >> beam.Map(lambda x: compute_stats(x[0], x[1]))\n",
      "  File \"/tmp/ipykernel_3858/3647296015.py\", line 3, in compute_stats\n",
      "    avg_arr_delay = float(np.mean(arrived)) if len(arrived) > 0 else None\n",
      "  File \"<__array_function__ internals>\", line 200, in mean\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/numpy/core/fromnumeric.py\", line 3464, in mean\n",
      "    return _methods._mean(a, axis=axis, dtype=dtype,\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/numpy/core/_methods.py\", line 194, in _mean\n",
      "    ret = ret / rcount\n",
      "TypeError: unsupported operand type(s) for /: 'NoneType' and 'int'\n",
      "ERROR:apache_beam.runners.direct.executor:Exception at bundle <apache_beam.runners.direct.bundle_factory._Bundle object at 0x78d32553f180>, due to an exception.\n",
      " Traceback (most recent call last):\n",
      "  File \"apache_beam/runners/common.py\", line 1434, in apache_beam.runners.common.DoFnRunner.process\n",
      "  File \"apache_beam/runners/common.py\", line 637, in apache_beam.runners.common.SimpleInvoker.invoke_process\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/apache_beam/transforms/core.py\", line 1963, in <lambda>\n",
      "    wrapper = lambda x: [fn(x)]\n",
      "  File \"/tmp/ipykernel_3858/2327646342.py\", line 21, in <lambda>\n",
      "    | 'stats' >> beam.Map(lambda x: compute_stats(x[0], x[1]))\n",
      "  File \"/tmp/ipykernel_3858/3647296015.py\", line 3, in compute_stats\n",
      "    avg_arr_delay = float(np.mean(arrived)) if len(arrived) > 0 else None\n",
      "  File \"<__array_function__ internals>\", line 200, in mean\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/numpy/core/fromnumeric.py\", line 3464, in mean\n",
      "    return _methods._mean(a, axis=axis, dtype=dtype,\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/numpy/core/_methods.py\", line 194, in _mean\n",
      "    ret = ret / rcount\n",
      "TypeError: unsupported operand type(s) for /: 'NoneType' and 'int'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/apache_beam/runners/direct/executor.py\", line 370, in call\n",
      "    self.attempt_call(\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/apache_beam/runners/direct/executor.py\", line 413, in attempt_call\n",
      "    evaluator.process_element(value)\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 883, in process_element\n",
      "    self.runner.process(element)\n",
      "  File \"apache_beam/runners/common.py\", line 1431, in apache_beam.runners.common.DoFnRunner.process\n",
      "  File \"apache_beam/runners/common.py\", line 1436, in apache_beam.runners.common.DoFnRunner.process\n",
      "  File \"apache_beam/runners/common.py\", line 1545, in apache_beam.runners.common.DoFnRunner._reraise_augmented\n",
      "  File \"apache_beam/runners/common.py\", line 1434, in apache_beam.runners.common.DoFnRunner.process\n",
      "  File \"apache_beam/runners/common.py\", line 637, in apache_beam.runners.common.SimpleInvoker.invoke_process\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/apache_beam/transforms/core.py\", line 1963, in <lambda>\n",
      "    wrapper = lambda x: [fn(x)]\n",
      "  File \"/tmp/ipykernel_3858/2327646342.py\", line 21, in <lambda>\n",
      "    | 'stats' >> beam.Map(lambda x: compute_stats(x[0], x[1]))\n",
      "  File \"/tmp/ipykernel_3858/3647296015.py\", line 3, in compute_stats\n",
      "    avg_arr_delay = float(np.mean(arrived)) if len(arrived) > 0 else None\n",
      "  File \"<__array_function__ internals>\", line 200, in mean\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/numpy/core/fromnumeric.py\", line 3464, in mean\n",
      "    return _methods._mean(a, axis=axis, dtype=dtype,\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/numpy/core/_methods.py\", line 194, in _mean\n",
      "    ret = ret / rcount\n",
      "TypeError: unsupported operand type(s) for /: 'NoneType' and 'int' [while running '[4]: stats']\n",
      "\n",
      "ERROR:apache_beam.runners.common:unsupported operand type(s) for /: 'NoneType' and 'int' [while running '[4]: stats']\n",
      "Traceback (most recent call last):\n",
      "  File \"apache_beam/runners/common.py\", line 1434, in apache_beam.runners.common.DoFnRunner.process\n",
      "  File \"apache_beam/runners/common.py\", line 637, in apache_beam.runners.common.SimpleInvoker.invoke_process\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/apache_beam/transforms/core.py\", line 1963, in <lambda>\n",
      "    wrapper = lambda x: [fn(x)]\n",
      "  File \"/tmp/ipykernel_3858/2327646342.py\", line 21, in <lambda>\n",
      "    | 'stats' >> beam.Map(lambda x: compute_stats(x[0], x[1]))\n",
      "  File \"/tmp/ipykernel_3858/3647296015.py\", line 3, in compute_stats\n",
      "    avg_arr_delay = float(np.mean(arrived)) if len(arrived) > 0 else None\n",
      "  File \"<__array_function__ internals>\", line 200, in mean\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/numpy/core/fromnumeric.py\", line 3464, in mean\n",
      "    return _methods._mean(a, axis=axis, dtype=dtype,\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/numpy/core/_methods.py\", line 194, in _mean\n",
      "    ret = ret / rcount\n",
      "TypeError: unsupported operand type(s) for /: 'NoneType' and 'int'\n",
      "ERROR:apache_beam.runners.direct.executor:Exception at bundle <apache_beam.runners.direct.bundle_factory._Bundle object at 0x78d32553f180>, due to an exception.\n",
      " Traceback (most recent call last):\n",
      "  File \"apache_beam/runners/common.py\", line 1434, in apache_beam.runners.common.DoFnRunner.process\n",
      "  File \"apache_beam/runners/common.py\", line 637, in apache_beam.runners.common.SimpleInvoker.invoke_process\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/apache_beam/transforms/core.py\", line 1963, in <lambda>\n",
      "    wrapper = lambda x: [fn(x)]\n",
      "  File \"/tmp/ipykernel_3858/2327646342.py\", line 21, in <lambda>\n",
      "    | 'stats' >> beam.Map(lambda x: compute_stats(x[0], x[1]))\n",
      "  File \"/tmp/ipykernel_3858/3647296015.py\", line 3, in compute_stats\n",
      "    avg_arr_delay = float(np.mean(arrived)) if len(arrived) > 0 else None\n",
      "  File \"<__array_function__ internals>\", line 200, in mean\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/numpy/core/fromnumeric.py\", line 3464, in mean\n",
      "    return _methods._mean(a, axis=axis, dtype=dtype,\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/numpy/core/_methods.py\", line 194, in _mean\n",
      "    ret = ret / rcount\n",
      "TypeError: unsupported operand type(s) for /: 'NoneType' and 'int'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/apache_beam/runners/direct/executor.py\", line 370, in call\n",
      "    self.attempt_call(\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/apache_beam/runners/direct/executor.py\", line 413, in attempt_call\n",
      "    evaluator.process_element(value)\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 883, in process_element\n",
      "    self.runner.process(element)\n",
      "  File \"apache_beam/runners/common.py\", line 1431, in apache_beam.runners.common.DoFnRunner.process\n",
      "  File \"apache_beam/runners/common.py\", line 1436, in apache_beam.runners.common.DoFnRunner.process\n",
      "  File \"apache_beam/runners/common.py\", line 1545, in apache_beam.runners.common.DoFnRunner._reraise_augmented\n",
      "  File \"apache_beam/runners/common.py\", line 1434, in apache_beam.runners.common.DoFnRunner.process\n",
      "  File \"apache_beam/runners/common.py\", line 637, in apache_beam.runners.common.SimpleInvoker.invoke_process\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/apache_beam/transforms/core.py\", line 1963, in <lambda>\n",
      "    wrapper = lambda x: [fn(x)]\n",
      "  File \"/tmp/ipykernel_3858/2327646342.py\", line 21, in <lambda>\n",
      "    | 'stats' >> beam.Map(lambda x: compute_stats(x[0], x[1]))\n",
      "  File \"/tmp/ipykernel_3858/3647296015.py\", line 3, in compute_stats\n",
      "    avg_arr_delay = float(np.mean(arrived)) if len(arrived) > 0 else None\n",
      "  File \"<__array_function__ internals>\", line 200, in mean\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/numpy/core/fromnumeric.py\", line 3464, in mean\n",
      "    return _methods._mean(a, axis=axis, dtype=dtype,\n",
      "  File \"/home/antoine_baret/data-science-on-gcp/.venv/lib/python3.9/site-packages/numpy/core/_methods.py\", line 194, in _mean\n",
      "    ret = ret / rcount\n",
      "TypeError: unsupported operand type(s) for /: 'NoneType' and 'int' [while running '[4]: stats']\n",
      "\n",
      "ERROR:apache_beam.runners.direct.executor:Giving up after 4 attempts.\n",
      "WARNING:apache_beam.runners.direct.executor:A task failed with exception: unsupported operand type(s) for /: 'NoneType' and 'int' [while running '[4]: stats']\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n",
      "WARNING:google.auth._default:No project ID could be determined. Consider running `gcloud config set project` or setting the GOOGLE_CLOUD_PROJECT environment variable\n"
     ]
    }
   ],
   "source": [
    "!{'gcloud config set project poc-dbs'}\n",
    "!{'export GOOGLE_CLOUD_PROJECT=poc-dbs'}\n",
    "with beam.Pipeline(argv=argv) as pipeline:\n",
    "        events = {}\n",
    "\n",
    "        for event_name in ['arrived', 'departed']:\n",
    "            topic_name = \"projects/{}/topics/{}\".format(project, event_name)\n",
    "\n",
    "            events[event_name] = (pipeline\n",
    "                                  | 'read:{}'.format(event_name) >> beam.io.ReadFromPubSub(\n",
    "                                                topic=topic_name)\n",
    "                                  | 'parse:{}'.format(event_name) >> beam.Map(lambda s: json.loads(s))\n",
    "                                  )\n",
    "\n",
    "        all_events = (events['arrived'], events['departed']) | beam.Flatten()\n",
    "\n",
    "        stats = (all_events\n",
    "                 | 'byairport' >> beam.Map(by_airport)\n",
    "                 | 'window' >> beam.WindowInto(beam.window.SlidingWindows(60 * 60, 5 * 60))\n",
    "                 | 'group' >> beam.GroupByKey()\n",
    "                 | 'stats' >> beam.Map(lambda x: compute_stats(x[0], x[1]))\n",
    "                 | beam.Map(lambda x: print(x))\n",
    "        )\n",
    "\n",
    "\n",
    "        \"\"\"\n",
    "        stats_schema = ','.join(['AIRPORT:string,AVG_ARR_DELAY:float,AVG_DEP_DELAY:float',\n",
    "                                 'NUM_FLIGHTS:int64,START_TIME:timestamp,END_TIME:timestamp']\n",
    "                                 )\n",
    "        (stats\n",
    "         | 'bqout' >> beam.io.gcp.bigquery.WriteToBigQuery(\n",
    "                    table='dsongcp.streaming_delays', \n",
    "                    schema=stats_schema\n",
    "                )\n",
    "         )\"\"\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
